MaxText/layers/linears.py

```diff
@@ -115,49 +167,96 @@ def __call__(self, inputs: Array) -> Array:
-      kernel = self.param(
-          "kernel",
-          nn.with_logical_partitioning(self.kernel_init, self.kernel_axes),
-          kernel_shape,
-          self.weight_dtype,
-          kernel_in_axis,
-          kernel_out_axis,
-      )
-    kernel = jnp.asarray(kernel, self.dtype)
-    contract_ind = tuple(range(0, len(axis)))
-    output = _compute_dot_general(inputs, kernel, self.kernel_axes, axis, contract_ind, self.matmul_precision, self.quant)
+      kernel = jnp.asarray(self.kernel[...], self.dtype)
+
+    contract_ind = tuple(range(0, len(self.axis)))
+    output = _compute_dot_general(
+        inputs,
+        kernel,
+        self.kernel_axes,
+        norm_axis,
+        contract_ind,
+        self.matmul_precision,
+        self.quant,
+    )
 
-    if self.use_bias:
-      bias_axes, bias_shape = (
-          self.kernel_axes[-len(features) :],
-          kernel_shape[-len(features) :],
-      )
-      bias = self.param(
-          "bias",
-          nn.with_logical_partitioning(bias_init, bias_axes),
-          bias_shape,
-          self.weight_dtype,
-      )
-      bias = jnp.asarray(bias, self.dtype)
+    if self.bias is not None:
+      bias = jnp.asarray(self.bias[...], self.dtype)
       output += bias
     return output
 
 
+def variable_to_logically_partitioned(variable: nnx.VariableState):
+  metadata = variable.get_metadata()
+  return nn.LogicallyPartitioned(  # type: ignore[wrong-keyword-args]
+      variable.value,
+      variable.sharding,
+      mesh=metadata.get("mesh"),
+      rules=metadata.get("rules"),
+  )
+
+
+def dense_general(
+    *,
+    inputs_shape: tuple[int, ...] | None = None,
+    in_features: tuple[int, ...] | int | None = None,
+    features: Union[Iterable[int], int],
+    axis: Union[Iterable[int], int] = -1,
+    weight_dtype: DType = jnp.float32,
+    dtype: DType = jnp.float32,
+    kernel_init: NdInitializer = nd_dense_init(
+        1.0, "fan_in", "truncated_normal"
+    ),
+    kernel_axes: Tuple[Optional[str], ...] = (),
+    quant: Optional[Quant] = None,
+    use_bias: bool = False,
+    matmul_precision: str = "default",
+    name: Optional[str] = None,
+):
+  if not (inputs_shape is not None) ^ (in_features is not None):
+    raise ValueError(
+        "Exactly one of inputs_shape or in_features must be specified."
+    )
+
+  if inputs_shape is not None:
+    axis = _canonicalize_tuple(axis)
+    in_features = tuple(
+        inputs_shape[ax] for ax in _normalize_axes(axis, len(inputs_shape))
+    )
+  else:
+    assert in_features is not None
+  module = nnx.bridge.to_linen(
+      DenseGeneral,
+      in_features=in_features,
+      out_features=features,
+      axis=axis,
+      weight_dtype=weight_dtype,
+      dtype=dtype,
+      kernel_init=kernel_init,
+      kernel_axes=kernel_axes,
+      quant=quant,
+      use_bias=use_bias,
+      matmul_precision=matmul_precision,
+      name=name,
+      metadata_fn=variable_to_logically_partitioned,
+  )
+  return module
```

Comment:
Just to confirm my understanding - this will not always be a bridge to linen, right? This is just a starting point for the migration and we will eventually have everything just be nnx

Reply:
correct!